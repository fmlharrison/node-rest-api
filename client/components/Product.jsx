import React from "react";
import PropTypes from "prop-types";

const Product = ({ info }) => {
  return (
    <div data-product-id={info.id}>
      <h4>{info.name}</h4>
      <p>{info.price}</p>
    </div>
  );
};

export default Product;

Product.propTypes = {
  info: PropTypes.shape({
    id: PropTypes.string,
    name: PropTypes.string,
    price: PropTypes.number
  }).isRequired
};
