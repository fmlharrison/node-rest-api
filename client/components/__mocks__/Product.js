export const product = {
  id: "12345",
  name: "Happy Meal",
  price: 5.99
};
