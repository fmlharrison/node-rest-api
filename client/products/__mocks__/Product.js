export const product = {
  id: "12345",
  name: "Happy Meal",
  price: 5.99
};

export const apiProductResponse = [
  {
    _id: "12345",
    name: "Happy Meal",
    price: 5.99
  }
];
