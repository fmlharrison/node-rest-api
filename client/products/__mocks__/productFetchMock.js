import { apiProductResponse } from "./Product";

// export default function apiGetMethod = jest.fn().mockImplementation(
//     new Promise((resolve, reject) => {
//         process.nextTick(
//             () => resolve(apiProductResponse);
//         );
//     });
// );
