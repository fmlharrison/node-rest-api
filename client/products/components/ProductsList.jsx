import React, { Component } from "react";
import Product from "./Product.jsx";

class ProductsList extends Component {
  constructor(props) {
    super(props);
    this.state = {
      products: []
    };
  }

  componentDidMount() {
    fetch("http://localhost:3000/products")
      .then(
        res => {
          return res.json();
        },
        err => {
          err.json();
        }
      )
      .then(data => {
        const products = data.map(item => {
          return {
            id: item._id,
            name: item.name,
            price: item.price
          };
        });
        this.setState({ products });
      });
  }

  render() {
    return (
      <div>
        <h3>Available Product</h3>
        {this.renderProductList()}
      </div>
    );
  }

  renderProductList = () => {
    const { products } = this.state;
    return (
      <ul style={{ paddingLeft: "0px" }}>
        {products.map(product => (
          <li key={product.id} style={{ listStyle: "none" }}>
            <Product info={product} />
          </li>
        ))}
      </ul>
    );
  };
}

export default ProductsList;
