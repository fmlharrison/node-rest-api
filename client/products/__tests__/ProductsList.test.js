import React from "react";
import sinon from "sinon";

import ProductsList from "../components/ProductsList.jsx";
import { product, apiProductResponse } from "../__mocks__/Product";
import fetchAPI from "../helpers/fetch";

describe("Shallow render of ProductList compoent", () => {
  it("Renders and empty product list", () => {
    const wrapper = shallow(<ProductsList />);
    expect(wrapper).toMatchSnapshot();
  });
});

describe("Render product list component", () => {
  it("Renders and fetches products", () => {
    const promise = Promise.resolve(apiProductResponse);
    sinon.stub(fetchAPI, "fetchProducts").returns(promise);

    const wrapper = mount(<ProductsList />);

    return promise
      .then(() => {
        expect(wrapper.state().products).toHaveLength(1);

        wrapper.update();
      })
      .then(() => {
        expect(wrapper).toMatchSnapshot();
      });
  });
});
