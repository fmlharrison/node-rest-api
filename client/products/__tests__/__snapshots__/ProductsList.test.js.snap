// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render product list component Renders and fetches products 1`] = `
<ProductsList>
  <div>
    <h3>
      Available Product
    </h3>
    <div>
      <ProductsButton>
        <button
          onClick={[Function]}
        >
          Show Products
        </button>
      </ProductsButton>
    </div>
    <ul
      style={
        Object {
          "paddingLeft": "0px",
        }
      }
    >
      <li
        key="12345"
        style={
          Object {
            "listStyle": "none",
          }
        }
      >
        <Product
          info={
            Object {
              "_id": "12345",
              "name": "Happy Meal",
              "price": 5.99,
            }
          }
        >
          <div
            data-product-id="12345"
          >
            <h4>
              Happy Meal
            </h4>
            <p>
              5.99
            </p>
          </div>
        </Product>
      </li>
    </ul>
  </div>
</ProductsList>
`;

exports[`Shallow render of ProductList compoent Renders and empty product list 1`] = `
<div>
  <h3>
    Available Product
  </h3>
  <div>
    <ProductsButton />
  </div>
  <ul
    style={
      Object {
        "paddingLeft": "0px",
      }
    }
  />
</div>
`;
